File: src/reader.rs
/** Serene --- Yet an other Lisp
*
* Copyright (c) 2020  Sameer Rahmani <lxsameer@gnu.org>
*
* This program is free software; you can redistribute it and/or modify
* it under the terms of the GNU General Public License as published by
* the Free Software Foundation, either version 2 of the License.
*
* This program is distributed in the hope that it will be useful,
* but WITHOUT ANY WARRANTY; without even the implied warranty of
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
* GNU General Public License for more details.
*
* You should have received a copy of the GNU General Public License
* along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/
use crate::ast::Expr;
use crate::types::Number;
use std::io::{BufReader, Read};

pub type ReadResult<'a> = Result<Expr, String>;

pub struct ExprReader {
    location: i32,
    read_stack: Vec<char>,
}

impl ExprReader {
    fn new() -> ExprReader {
        ExprReader {
            location: 0,
            read_stack: vec![],
        }
    }

    /// Retun a single character by reading from the `reader`. ,
    ///
    /// # Arguments:
    ///
    /// * `reader`: The buffer to read from.
    /// * `skip_whitespace`: Whether or not to skip whitespace chars. *Bear in mind that
    ///   if you care about the newline char you should not skip the whitespaces*.
    fn get_char<T: Read>(
        &mut self,
        reader: &mut BufReader<T>,
        skip_whitespace: bool,
    ) -> Option<char> {
        loop {
            match self.read_stack.pop() {
                Some(c) if !c.is_whitespace() || !skip_whitespace => return Some(c),
                Some(_) => continue,
                None => (),
            };

            // Rust is weird, it doesn't provide a way to read from a buffer char by char.
            let mut single_char_buff = [0];
            let bytes_read = reader.read(&mut single_char_buff);
            match bytes_read {
                Ok(n) if n > 0 => self.location = self.location + 1,
                Ok(_) => return None,
                Err(_) => return None,
            };
            let ch = single_char_buff[0] as char;

            match ch {
                c if !c.is_whitespace() || !skip_whitespace => return Some(c),
                _ => (),
            };
        }
    }

    fn unget_char(&mut self, c: char) {
        self.read_stack.push(c);
    }

    // Look ahead. AFAIK Rust doesn't provide any unread functoinality like Java input streams which
    // sucks.
    fn peek_char<T: Read>(
        &mut self,
        reader: &mut BufReader<T>,
        skip_whitespace: bool,
    ) -> Option<char> {
        match self.get_char(reader, skip_whitespace) {
            Some(c) => {
                self.unget_char(c);
                Some(c)
            }
            None => None,
        }
    }

    fn read_quoted_expr<T: Read>(&mut self, reader: &mut BufReader<T>) -> ReadResult {
        let rest = self.read_expr(reader)?;
        let elements = vec![Expr::make_symbol("quote".to_string()), rest];

        Ok(Expr::make_list(&elements))
    }

    fn read_unquoted_expr<T: Read>(&mut self, reader: &mut BufReader<T>) -> ReadResult {
        match self.peek_char(reader, true) {
            Some('@') => {
                // Move forward in the buffer since we peeked it
                let _ = self.get_char(reader, true);
                let rest = self.read_expr(reader)?;
                let elements = vec![Expr::make_symbol("unquote-splicing".to_string()), rest];
                Ok(Expr::make_list(&elements))
            }
            _ => {
                let rest = self.read_expr(reader)?;
                let elements = vec![Expr::make_symbol("unquote".to_string()), rest];
                Ok(Expr::make_list(&elements))
            }
        }
    }

    fn read_quasiquoted_expr<T: Read>(&mut self, reader: &mut BufReader<T>) -> ReadResult {
        let rest = self.read_expr(reader)?;
        let elements = vec![Expr::make_symbol("quasiquote".to_string()), rest];
        Ok(Expr::make_list(&elements))
    }

    // TODO: We might want to replace Cons with an actual List struct
    fn read_list<T: Read>(&mut self, reader: &mut BufReader<T>) -> ReadResult {
        let mut result = Expr::make_empty_list();

        match self.read_expr(reader) {
            Ok(value) => result.push(value),
            Err(e) => match self.get_char(reader, true) {
                Some(')') => return Ok(Expr::list_to_cons(result)),
                _ => return Err(e),
            },
        };

        loop {
            match self.get_char(reader, true) {
                Some(')') => return Ok(Expr::list_to_cons(result)),
                Some(e) => {
                    self.unget_char(e);
                    result.push(self.read_expr(reader)?)
                }
                None => return Err("Unexpected EOF while parsing a list.".to_string()),
            };
        }
    }

    fn is_valid_for_identifier(&self, c: char) -> bool {
        match c {
            '!'
            | '$'
            | '%'
            | '&'
            | '*'
            | '+'
            | '-'
            | '.'
            | '~'
            | '/'
            | ':'
            | '<'
            | '='
            | '>'
            | '?'
            | '@'
            | '^'
            | '_'
            | 'a'..='z'
            | 'A'..='Z'
            | '0'..='9' => true,
            _ => false,
        }
    }

    fn _read_symbol<T: Read>(&mut self, reader: &mut BufReader<T>) -> ReadResult {
        let mut symbol = match self.peek_char(reader, false) {
            Some(e) if self.is_valid_for_identifier(e) => {
                // Read into string
                let ch = self.get_char(reader, false).unwrap();
                let mut s = String::new();
                s.push(ch);
                s
            }
            Some(e) => {
                return Err(format!(
                    "Unexpected character: got '{}', expected a symbol at {}",
                    e, self.location
                ))
            }
            None => return Err("Unexpected EOF".to_string()),
        };

        loop {
            match self.get_char(reader, false) {
                Some(v) if self.is_valid_for_identifier(v) => symbol.push(v),
                Some(v) => {
                    self.unget_char(v);
                    break;
                }
                None => break,
            }
        }

        Ok(Expr::make_symbol(symbol))
    }

    fn read_escape_char<T: Read>(&mut self, reader: &mut BufReader<T>) -> Option<char> {
        match self.get_char(reader, false) {
            Some(e) => match e {
                '\"' => Some('\"'),
                '\'' => Some('\''),
                '\\' => Some('\\'),
                'n' => Some('\n'),
                'r' => Some('\r'),
                't' => Some('\t'),
                _ => None,
            },
            None => None,
        }
    }

    fn read_string<T: Read>(&mut self, reader: &mut BufReader<T>) -> ReadResult {
        // Skip the " char
        let _ = self.get_char(reader, false);

        let mut string = "".to_string();
        loop {
            match self.get_char(reader, false) {
                Some(e) => match e {
                    '\"' => break,
                    '\\' => match self.read_escape_char(reader) {
                        Some(v) => string.push(v),
                        None => return Err(format!("Unexpected char to escape, got {}", e)),
                    },
                    //'\n' => return Err("Unescaped newlines are not allowed in string literals".to_string()),
                    _ => string.push(e),
                },
                None => return Err("Unexpected EOF while scanning a string".to_string()),
            }
        }
        Ok(Expr::make_string(string))
    }

    fn read_number<T: Read>(&mut self, reader: &mut BufReader<T>, neg: bool) -> ReadResult {
        let mut is_double = false;
        let mut string = (if neg { "-" } else { "" }).to_string();

        loop {
            match self.get_char(reader, false) {
                Some(e) if e == '.' && is_double => {
                    return Err("A double with more that one '.' ???".to_string())
                }
                Some(e) if e == '.' => {
                    is_double = true;
                    string.push(e);
                }
                Some(e) if e.is_digit(10) => string.push(e),
                Some(e) => {
                    self.unget_char(e);
                    break;
                }
                None => break,
            }
        }

        // TODO: Move this to ast module and use the `new` function on
        // Number struct
        if is_double {
            Ok(Expr::make_number(Number::Float(
                string.parse::<f64>().unwrap(),
            )))
        } else {
            Ok(Expr::make_number(Number::Integer(
                string.parse::<i64>().unwrap(),
            )))
        }
    }

    fn read_symbol<T: Read>(&mut self, reader: &mut BufReader<T>) -> ReadResult {
        match self.peek_char(reader, true) {
            Some(c) => match c {
                '\"' => self.read_string(reader),
                c if c.is_digit(10) => self.read_number(reader, false),

                // ':' => self.read_keyword(reader),
                '-' => {
                    // Read the '-' char
                    let _ = self.get_char(reader, true);
                    match self.peek_char(reader, true) {
                        Some(ch) => match ch {
                            ch if ch.is_digit(10) => self.read_number(reader, true),
                            _ => {
                                self.unget_char(c);
                                self._read_symbol(reader)
                            }
                        },
                        None => {
                            self.unget_char(c);
                            self._read_symbol(reader)
                        }
                    }
                }

                _ => self._read_symbol(reader),
            },
            None => Err("Unexpected EOF while scanning atom".to_string()),
        }
    }

    pub fn ignore_comments<T: Read>(&mut self, reader: &mut BufReader<T>) -> ReadResult {
        match self.get_char(reader, false) {
            Some(c) => match c {
                '\n' => Ok(Expr::Comment),
                _ => self.ignore_comments(reader),
            },
            None => Ok(Expr::Comment),
        }
    }

    pub fn read_expr<T: Read>(&mut self, reader: &mut BufReader<T>) -> ReadResult {
        match self.get_char(reader, true) {
            Some(c) => {
                match c {
                    '\'' => self.read_quoted_expr(reader),
                    '~' => self.read_unquoted_expr(reader),
                    '`' => self.read_quasiquoted_expr(reader),
                    '(' => self.read_list(reader),
                    ';' => self.ignore_comments(reader),
                    //'[' => self.read_vector(reader),
                    //'{' => self.read_map(reader),
                    _ => {
                        self.unget_char(c);
                        self.read_symbol(reader)
                    }
                }
            }
            None => Ok(Expr::NoMatch),
        }
    }

    pub fn read_from_buffer<T: Read>(
        &mut self,
        reader: &mut BufReader<T>,
    ) -> Result<Vec<Expr>, String> {
        let mut ast = vec![];
        loop {
            match self.read_expr(reader) {
                Ok(Expr::NoMatch) => break,
                Err(v) => return Err(v),
                Ok(Expr::Comment) => continue,
                Ok(v) => ast.push(v),
            }
        }
        Ok(ast)
    }

    pub fn read(&mut self, string: &str) -> Result<Vec<Expr>, String> {
        let reader = BufReader::new(string.as_bytes());
        let mut buf_reader = BufReader::new(reader);
        self.read_from_buffer(&mut buf_reader)
    }
}

pub fn read_string(input: &str) -> Result<Vec<Expr>, String> {
    let mut reader = ExprReader::new();
    reader.read(input)
}

File: src/libserene/reader/reader.cpp
/* -*- C++ -*-
 * Serene Programming Language
 *
 * Copyright (c) 2019-2021 Sameer Rahmani <lxsameer@gnu.org>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, version 2.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#include "serene/reader/reader.h"

#include "serene/errors/constants.h"
#include "serene/exprs/list.h"
#include "serene/exprs/number.h"
#include "serene/exprs/symbol.h"
#include "serene/namespace.h"

#include <assert.h>
#include <cctype>
#include <fstream>
#include <llvm/ADT/StringRef.h>
#include <llvm/Support/Error.h>
#include <llvm/Support/ErrorHandling.h>
#include <llvm/Support/ErrorOr.h>
#include <llvm/Support/FormatVariadic.h>
#include <llvm/Support/MemoryBuffer.h>
#include <llvm/Support/SMLoc.h>
#include <memory>
#include <mlir/IR/Diagnostics.h>
#include <mlir/IR/Location.h>
#include <mlir/Support/LogicalResult.h>
#include <string>

namespace serene {

namespace reader {

Reader::Reader(SereneContext &ctx, llvm::StringRef buffer, llvm::StringRef ns,
               llvm::Optional<llvm::StringRef> filename)
    : ctx(ctx), ns(ns), filename(filename), buf(buffer),
      currentLocation(Location(ns, filename)) {
  READER_LOG("Setting the first char of the buffer");
  currentChar          = buf.begin() - 1;
  currentPos           = 1;
  currentLocation.line = 1;
  currentLocation.col  = 1;
};

Reader::Reader(SereneContext &ctx, llvm::MemoryBufferRef buffer,
               llvm::StringRef ns, llvm::Optional<llvm::StringRef> filename)
    : Reader(ctx, buffer.getBuffer(), ns, filename){};

Reader::~Reader() { READER_LOG("Destroying the reader"); }

void Reader::advanceByOne() {
  currentChar++;
  currentPos++;
  currentLocation.col++;

  if (*currentChar == '\n') {
    READER_LOG("Detected end of line");

    if (readEOL) {
      currentLocation.col = 1;
      currentLocation.line++;
    }

    readEOL = true;
  } else {
    if (readEOL) {
      currentLocation.line++;
      currentLocation.col = 1;
    }
    readEOL = false;
  }

  READER_LOG("Moving to Char: " << *currentChar << " at location: "
                                << currentLocation.toString());
};
void Reader::advance(bool skipWhitespace) {
  if (skipWhitespace) {
    for (;;) {
      auto next = currentChar + 1;

      if (!isspace(*next)) {
        return;
      }

      advanceByOne();
    }
  } else {
    advanceByOne();
  }
};

const char *Reader::nextChar(bool skipWhitespace, unsigned count) {
  if (!skipWhitespace) {
    READER_LOG("Next char: " << *(currentChar + count));
    return currentChar + count;
  }

  auto c = currentChar + 1;
  while (isspace(*c)) {
    c++;
  };

  READER_LOG("Next char: " << *c);
  return c;
};

bool Reader::isEndOfBuffer(const char *c) {
  return *c == '\0' || currentPos > buf.size() || *c == EOF;
};

Location Reader::getCurrentLocation() { return currentLocation.clone(); };

/// A predicate function indicating whether the given char `c` is a valid
/// char for the starting point of a symbol or not.
bool Reader::isValidForIdentifier(char c) {
  switch (c) {
  case '!':
  case '$':
  case '%':
  case '&':
  case '*':
  case '+':
  case '-':
  case '.':
  case '~':
  case '/':
  case ':':
  case '<':
  case '=':
  case '>':
  case '?':
  case '@':
  case '^':
  case '_':
    return true;
  }

  if (std::isalnum(c)) {
    return true;
  }
  return false;
}

/// Reads a number,
/// \param neg whether to read a negative number or not.
exprs::Node Reader::readNumber(bool neg) {
  READER_LOG("Reading a number...");
  std::string number(neg ? "-" : "");
  bool floatNum = false;
  bool empty    = false;

  auto c = nextChar();
  advance();

  LocationRange loc(getCurrentLocation());

  if (!isdigit(*c)) {
    ctx.diagEngine->emitSyntaxError(loc, errors::InvalidDigitForNumber);
    exit(1);
  }

  for (;;) {
    number += *c;
    c     = nextChar(false);
    empty = false;

    if (isdigit(*c) || *c == '.') {
      if (*c == '.' && floatNum == true) {
        loc = LocationRange(getCurrentLocation());
        ctx.diagEngine->emitSyntaxError(loc, errors::TwoFloatPoints);
        exit(1);
      }

      if (*c == '.') {
        floatNum = true;
      }

      advance();
      continue;
    }
    break;
  }

  if ((std::isalpha(*c) && !empty) || empty) {
    advance();
    loc.start = getCurrentLocation();
    ctx.diagEngine->emitSyntaxError(loc, errors::InvalidDigitForNumber);
    exit(1);
  }

  loc.end = getCurrentLocation();
  return exprs::make<exprs::Number>(loc, number, neg, floatNum);
};

/// Reads a symbol. If the symbol looks like a number
/// If reads it as number
exprs::Node Reader::readSymbol() {
  READER_LOG("Reading a symbol...");
  LocationRange loc;
  auto c = nextChar();

  if (!this->isValidForIdentifier(*c) || isEndOfBuffer(c) || isspace(*c)) {
    advance();
    loc = LocationRange(getCurrentLocation());
    std::string msg;

    if (*c == ')') {
      msg = "An extra ')' is detected.";
    }

    ctx.diagEngine->emitSyntaxError(loc, errors::InvalidCharacterForSymbol,
                                    msg);
    exit(1);
  }

  if (*c == '-') {
    auto next = nextChar(false, 2);
    if (isdigit(*next)) {
      // Swallow the -
      advance();
      return readNumber(true);
    }
  }

  if (isdigit(*c)) {
    return readNumber(false);
  }

  std::string sym("");
  advance();

  for (;;) {
    sym += *c;
    c = nextChar();

    if (!isEndOfBuffer(c) &&
        ((!(isspace(*c)) && this->isValidForIdentifier(*c)))) {
      advance();
      continue;
    }
    break;
  }

  loc.end = getCurrentLocation();
  return exprs::make<exprs::Symbol>(loc, sym);
};

/// Reads a list recursively
exprs::Node Reader::readList() {
  READER_LOG("Reading a list...");

  auto c = nextChar();
  advance();

  auto list = exprs::makeAndCast<exprs::List>(getCurrentLocation());

  // TODO: Replace the assert with an actual check.
  assert(*c == '(');

  bool list_terminated = false;

  do {
    auto c = nextChar(true);

    if (isEndOfBuffer(c)) {
      advance(true);
      advance();
      list->location.end = getCurrentLocation();
      ctx.diagEngine->emitSyntaxError(list->location,
                                      errors::EOFWhileScaningAList);
      exit(1);
    }

    switch (*c) {
    case ')':
      advance(true);
      advance();
      list_terminated    = true;
      list->location.end = getCurrentLocation();
      break;

    default:
      advance(true);
      list->append(readExpr());
    }

  } while (!list_terminated);

  return list;
};

/// Reads an expression by dispatching to the proper reader function.
exprs::Node Reader::readExpr() {
  auto c = nextChar(true);

  READER_LOG("Read char at `readExpr`: " << *c);

  if (isEndOfBuffer(c)) {
    return nullptr;
  }

  switch (*c) {
  case '(': {
    advance(true);
    return readList();
  }

  default:
    advance(true);
    return readSymbol();
  }
};

/// Reads all the expressions in the reader's buffer as an AST.
/// Each expression type (from the reader perspective) has a
/// reader function.
Result<exprs::Ast> Reader::read() {

  for (size_t current_pos = 0; current_pos < buf.size();) {
    auto c = nextChar(true);

    if (isEndOfBuffer(c)) {
      break;
    }

    advance(true);

    auto tmp{readExpr()};

    if (tmp) {
      this->ast.push_back(move(tmp));
    } else {
      break;
    }
  }

  return Result<exprs::Ast>::success(std::move(this->ast));
};

Result<exprs::Ast> read(SereneContext &ctx, const llvm::StringRef input,
                        llvm::StringRef ns,
                        llvm::Optional<llvm::StringRef> filename) {
  reader::Reader r(ctx, input, ns, filename);
  auto ast = r.read();
  return ast;
}

Result<exprs::Ast> read(SereneContext &ctx, const llvm::MemoryBufferRef input,
                        llvm::StringRef ns,
                        llvm::Optional<llvm::StringRef> filename) {
  reader::Reader r(ctx, input, ns, filename);

  auto ast = r.read();
  return ast;
}
} // namespace reader
} // namespace serene